### PR TITLE
CLOUDSTACK-9806: Nuage domain template selection per VPC

### PR DESCRIPTION
**Pull request by prashanthvarma**
Co-Authored-By: Prashanth Manthena <prashanth.manthena@nuagenetworks.net>
Co-Authored-By: Frank Maximus <frank.maximus@nuagenetworks.net>

### CODE REVIEW

**Comment by prashanthvarma** [comment]
Design document for this feature: https://cwiki.apache.org/confluence/display/CLOUDSTACK/Nuage+domain+template+selection+per+VPC

**Comment by prashanthvarma** [comment]
PEP8 & PyFlakes compliance of our marvin test code:

CloudStack$
CloudStack$ pep8 test/integration/plugins/nuagevsp/.py
CloudStack$
CloudStack$ pyflakes test/integration/plugins/nuagevsp/.py
CloudStack$

Validations:

Marvin test results:

nosetests --with-marvin --marvin-config=nuage.cfg nuagevsp/test_nuage_vsp_domain_template.py

Test Nuage VSP Domain Template selection per VPC ... === TestName: test_01_nuage_Domain_Template_selection_per_VPC | Status : SUCCESS ===
ok
Test Nuage VSP Domain Template selection per VPC as ROOT domain ... === TestName: test_02_nuage_Domain_Template_selection_per_VPC_as_ROOT_user | Status : SUCCESS ===
ok
Test Nuage VSP Domain Template selection per VPC as domain admin ... === TestName: test_03_nuage_Domain_Template_selection_per_VPC_as_domain_admin | Status : SUCCESS ===
ok
Test Nuage VSP Domain Template selection per VPC as domain ... === TestName: test_04_nuage_Domain_Template_selection_per_VPC_as_domain_user | Status : SUCCESS ===
ok
Test Nuage VSP Domain Template selection per VPC as subdomain admin ... === TestName: test_05_nuage_Domain_Template_selection_per_VPC_as_subdom_admin | Status : SUCCESS ===
ok
Test Nuage VSP Domain Template selection per VPC as subdomain ... === TestName: test_06_nuage_Domain_Template_selection_per_VPC_as_subdom_user | Status : SUCCESS ===
ok
Test Nuage VSP Global Domain Template ... === TestName: test_07_nuage_Global_Domain_Template | Status : SUCCESS ===
ok
Test Nuage VSP Global Domain Template as ROOT domain regular user ... === TestName: test_08_nuage_Global_Domain_Template_as_ROOT_user | Status : SUCCESS ===
ok
Test Nuage VSP Global Domain Template as domain admin user ... === TestName: test_09_nuage_Global_Domain_Template_as_domain_admin | Status : SUCCESS ===
ok
Test Nuage VSP Global Domain Template as domain regular user ... === TestName: test_10_nuage_Global_Domain_Template_as_domain_user | Status : SUCCESS ===
ok
Test Nuage VSP Global Domain Template as subdomain admin user ... === TestName: test_11_nuage_Global_Domain_Template_as_subdomain_admin | Status : SUCCESS ===
ok
Test Nuage VSP Global Domain Template as subdomain regular user ... === TestName: test_12_nuage_Global_Domain_Template_as_subdomain_user | Status : SUCCESS ===
ok

----------------------------------------------------------------------
Ran 12 tests in 6203.613s

OK


[results.txt](https://github.com/apache/cloudstack/files/810482/results.txt)
[runinfo.txt](https://github.com/apache/cloudstack/files/810483/runinfo.txt)

**Comment by borisstoyanov** [comment]
Thanks for this good looking PR @prashanthvarma, it's really great that you provided integration tests. I'll package and kick regression tests to check if everything else is working fine with these changes.
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by prashanthvarma** [comment]
@borisstoyanov I am glad you liked the PR, we (@nuagenetworks) made it a practice to add and run integration tests that test the changes we propose/make in a PR. Moreover, we internally qualify through regressions the whole system with those changes for both Nuage VSP and native ACS setups/environments where necessary.    

Thank you for kicking the regression tests. Please do also code review the PR.

**Comment by borisstoyanov** [comment]
Thanks, the tests failures looks to be not related to these changes. @prashanthvarma

**Comment by prashanthvarma** [comment]
@borisstoyanov yes, I also did cross check the test failures. Thanks for the investigation.

**Comment by cloudmonger** [comment]
### ACS CI BVT Run
 **Sumarry:**
 Build Number 762
 Hypervisor xenserver
 NetworkType Advanced
 Passed=111
 Failed=1
 Skipped=12

_Link to logs Folder (search by build_no):_ https://www.dropbox.com/sh/r2si930m8xxzavs/AAAzNrnoF1fC3auFrvsKo_8-a?dl=0

**Failed tests:**
* test_non_contigiousvlan.py

 * test_extendPhysicalNetworkVlan Failed


**Skipped tests:**
test_vm_nic_adapter_vmxnet3
test_01_verify_libvirt
test_02_verify_libvirt_after_restart
test_03_verify_libvirt_attach_disk
test_04_verify_guest_lspci
test_05_change_vm_ostype_restart
test_06_verify_guest_lspci_again
test_static_role_account_acls
test_11_ss_nfs_version_on_ssvm
test_nested_virtualization_vmware
test_3d_gpu_support
test_deploy_vgpu_enabled_vm

**Passed test suits:**
test_deploy_vm_with_userdata.py
test_affinity_groups_projects.py
test_portable_publicip.py
test_vm_snapshots.py
test_over_provisioning.py
test_global_settings.py
test_scale_vm.py
test_service_offerings.py
test_routers_iptables_default_policy.py
test_loadbalance.py
test_routers.py
test_reset_vm_on_reboot.py
test_deploy_vms_with_varied_deploymentplanners.py
test_network.py
test_router_dns.py
test_login.py
test_deploy_vm_iso.py
test_list_ids_parameter.py
test_public_ip_range.py
test_multipleips_per_nic.py
test_metrics_api.py
test_regions.py
test_affinity_groups.py
test_network_acl.py
test_pvlan.py
test_volumes.py
test_nic.py
test_deploy_vm_root_resize.py
test_resource_detail.py
test_secondary_storage.py
test_vm_life_cycle.py
test_routers_network_ops.py
test_disk_offerings.py

**Comment by rhtyd** [comment]
@prashanthvarma can you see and fix CI failures?

**Comment by sgoeminn** [comment]
@rhtyd I rebased the PR, CI failure is a know issue (test_02_project_limits_normal_user) everything else looks good.

**Comment by fmaximus** [comment]
retest this please

**Comment by rhtyd** [comment]
@blueorangutan package

File: plugins/network-elements/nuage-vsp/src/com/cloud/api/commands/ListNuageVspDomainTemplatesCmd.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+import org.apache.cloudstack.api.response.ListResponse;
+import org.apache.cloudstack.api.response.PhysicalNetworkResponse;
+import org.apache.cloudstack.api.response.ZoneResponse;
+import org.apache.log4j.Logger;
+
+import javax.inject.Inject;
+import java.util.List;
+
+@APICommand(name = "listNuageVspDomainTemplates", responseObject = BaseResponse.class, description = "Lists Nuage VSP domain templates", authorized = {RoleType.Admin, RoleType.DomainAdmin, RoleType.User})
+public class ListNuageVspDomainTemplatesCmd extends BaseListCmd {
```
Comment (rhtyd) [inline plugins/network-elements/nuage-vsp/src/com/cloud/api/commands/ListNuageVspDomainTemplatesCmd.java]: @prashanthvarma have a look at ListRolesCmd and other newer implemented/added API for refering to a more modern CloudStack API implementation pattern.

**Comment by rhtyd** [comment]
@blueorangutan test

File: plugins/network-elements/nuage-vsp/src/com/cloud/api/commands/ListNuageVspDomainTemplatesCmd.java (hunk 1 above)
Comment (sgoeminn) [reply plugins/network-elements/nuage-vsp/src/com/cloud/api/commands/ListNuageVspDomainTemplatesCmd.java]: @rhtyd I made it aligned with the ListRolesCmd implementation.

**Comment by smeetsr** [comment]
After above commit , again I did a rerun of test_nuage_vpc_domain_template.py and all tests were successful.
Test Nuage VSP Domain Template selection per VPC ... === TestName: test_01_nuage_Domain_Template_selection_per_VPC | Status : SUCCESS ===
ok
Test Nuage VSP Domain Template selection per VPC as ROOT domain ... === TestName: test_02_nuage_Domain_Template_selection_per_VPC_as_ROOT_user | Status : SUCCESS ===
ok
Test Nuage VSP Domain Template selection per VPC as domain admin ... === TestName: test_03_nuage_Domain_Template_selection_per_VPC_as_domain_admin | Status : SUCCESS ===
ok
Test Nuage VSP Domain Template selection per VPC as domain ... === TestName: test_04_nuage_Domain_Template_selection_per_VPC_as_domain_user | Status : SUCCESS ===
ok
Test Nuage VSP Domain Template selection per VPC as subdomain admin ... === TestName: test_05_nuage_Domain_Template_selection_per_VPC_as_subdom_admin | Status : SUCCESS ===
ok
Test Nuage VSP Domain Template selection per VPC as subdomain ... === TestName: test_06_nuage_Domain_Template_selection_per_VPC_as_subdom_user | Status : SUCCESS ===
ok
Test Nuage VSP Global Domain Template ... === TestName: test_07_nuage_Global_Domain_Template | Status : SUCCESS ===
ok
Test Nuage VSP Global Domain Template as ROOT domain regular user ... === TestName: test_08_nuage_Global_Domain_Template_as_ROOT_user | Status : SUCCESS ===
ok
Test Nuage VSP Global Domain Template as domain admin user ... === TestName: test_09_nuage_Global_Domain_Template_as_domain_admin | Status : SUCCESS ===
ok
Test Nuage VSP Global Domain Template as domain regular user ... === TestName: test_10_nuage_Global_Domain_Template_as_domain_user | Status : SUCCESS ===
ok
Test Nuage VSP Global Domain Template as subdomain admin user ... === TestName: test_11_nuage_Global_Domain_Template_as_subdomain_admin | Status : SUCCESS ===
ok
Test Nuage VSP Global Domain Template as subdomain regular user ... === TestName: test_12_nuage_Global_Domain_Template_as_subdomain_user | Status : SUCCESS ===
ok

----------------------------------------------------------------------
Ran 12 tests in 5290.439s

OK

File: plugins/network-elements/nuage-vsp/src/com/cloud/api/commands/AssociateNuageVspDomainTemplateCmd.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+import org.apache.cloudstack.api.Parameter;
+import org.apache.cloudstack.api.ServerApiException;
+import org.apache.cloudstack.api.response.PhysicalNetworkResponse;
+import org.apache.cloudstack.api.response.SuccessResponse;
+import org.apache.cloudstack.api.response.VpcResponse;
+import org.apache.cloudstack.api.response.ZoneResponse;
+import org.apache.cloudstack.context.CallContext;
+import org.apache.log4j.Logger;
+
+        import javax.inject.Inject;
```
Comment (rhtyd) [inline plugins/network-elements/nuage-vsp/src/com/cloud/api/commands/AssociateNuageVspDomainTemplateCmd.java]: @sgoeminn fix indent please.

File: plugins/network-elements/nuage-vsp/src/com/cloud/api/commands/AssociateNuageVspDomainTemplateCmd.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+@APICommand(name = "associateNuageVspDomainTemplate", responseObject = SuccessResponse.class, description = "associate a vpc with a domain template", authorized = {RoleType.Admin, RoleType.DomainAdmin, RoleType.User})
+public class AssociateNuageVspDomainTemplateCmd extends BaseCmd {
+    private static final Logger s_logger = Logger.getLogger(AssociateNuageVspDomainTemplateCmd.class.getName());
+    private static final String s_name = "associatenuagevspdomaintemplateresponse";
```
Comment (rhtyd) [inline plugins/network-elements/nuage-vsp/src/com/cloud/api/commands/AssociateNuageVspDomainTemplateCmd.java]: @sgoeminn this still does not declare use `APINAME` as a reference/usage pattern as seen in recent/modern API implementations such as ListRolesCmd etc.

File: plugins/network-elements/nuage-vsp/src/com/cloud/api/commands/ListNuageVspDomainTemplatesCmd.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+import org.apache.cloudstack.api.response.DomainResponse;
+import org.apache.cloudstack.api.response.ListResponse;
+import org.apache.cloudstack.api.response.PhysicalNetworkResponse;
+import org.apache.cloudstack.api.response.ZoneResponse;
+
+import javax.inject.Inject;
+import java.util.List;
+@APICommand(name = ListNuageVspDomainTemplatesCmd.APINAME, responseObject = BaseResponse.class, description = "Lists Nuage VSP domain templates", since = "4.11", responseHasSensitiveInfo = false, authorized = {RoleType.Admin, RoleType.DomainAdmin, RoleType.User})
+public class ListNuageVspDomainTemplatesCmd extends BaseCmd {
+    static final String APINAME = "listNuageVspDomainTemplates";
```
Comment (rhtyd) [inline plugins/network-elements/nuage-vsp/src/com/cloud/api/commands/ListNuageVspDomainTemplatesCmd.java]: @sgoeminn thanks, LGTM. Please follow the same across other Cmd implementations.

File: plugins/network-elements/nuage-vsp/src/com/cloud/api/commands/ListNuageVspDomainTemplatesCmd.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+import org.apache.cloudstack.api.BaseResponse;
+import org.apache.cloudstack.api.Parameter;
+import org.apache.cloudstack.api.ServerApiException;
+import org.apache.cloudstack.api.response.DomainResponse;
+import org.apache.cloudstack.api.response.ListResponse;
+import org.apache.cloudstack.api.response.PhysicalNetworkResponse;
+import org.apache.cloudstack.api.response.ZoneResponse;
+
+import javax.inject.Inject;
+import java.util.List;
```
Comment (rhtyd) [inline plugins/network-elements/nuage-vsp/src/com/cloud/api/commands/ListNuageVspDomainTemplatesCmd.java]: @sgoeminn newline here please, thanks.

File: plugins/network-elements/nuage-vsp/src/com/cloud/api/commands/ListNuageVspDomainTemplatesCmd.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    public Long getDomainId() {
+        return domainId;
+    }
+
+    public Long getPhysicalNetworkId() {
+        return physicalNetworkId;
+    }
+
+    public Long getZoneId() { return zoneId; }
+/////////////////////////////////////////////////////
```
Comment (rhtyd) [inline plugins/network-elements/nuage-vsp/src/com/cloud/api/commands/ListNuageVspDomainTemplatesCmd.java]: @sgoeminn fix indents and newlines please.

**Comment by sgoeminn** [comment]
@rhtyd Thanks for reviewing! I addressed all your comments.

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@fmaximus thanks for merging, kindly squash merge or rebase/merge in future.

**Pull request merged**